*** Modules/makesetup.orig	2019-10-17 14:58:13.484348696 -0700
--- Modules/makesetup	2019-10-17 14:59:05.972375284 -0700
***************
*** 133,138 ****
--- 133,139 ----
  
  		# Output DEFS in reverse order so first definition overrides
  		case $line in
+ 		*-D*) ;;  # -D can include '='
  		*=*)	DEFS="$line$NL$DEFS"; continue;;
  		'include '*)	DEFS="$line$NL$DEFS"; continue;;
  		'*noobjects*')
